### PR TITLE
[UI] Move Runtime status to BottomAppBar

### PR DESCRIPTION
**Pull request by senwang86**
## Summary
A UI refactor which moves the `Sync Status` and `Runtime Status`  from `Sidebar` to a newly created `BottomAppBar`.

Screenshot 2023-08-29 at 4 23 40 PM

## Test

  
bottom_status_bar

### CODE REVIEW

File: apps/ui/src/pages/repo.tsx [hunk 1]
```diff
@@ -215,14 +216,19 @@ function RepoWrapper({ children, id }) {
           sx={{
             boxSizing: "border-box",
             width: "100%",
-            height: "100%",
+            height: "96%",
```
Comment (lihebi) [inline apps/ui/src/pages/repo.tsx:219]: This looks fragile.

**Comment by lihebi** [comment]
This UI looks fancy! Instead of "moving", I'd like to keep both (essentially two types of "widgets"). We need a "widget" system to allow users to configure what widgets to show in the sidebar and the bottom-status-bar.

**Comment by lihebi** [comment]
I changed the sidebar just a little bit in #493, resulting in a slight conflict. I've merged that PR since it's pretty big and may cause more conflicts.

**Comment by lihebi** [comment]
A second thought: I probably want to hold on this, because the status items/buttons aren’t fixed, and I’m working on adding locally spawned runtimes to these sections. Working with sidebar seems easier to make changes.

**Comment by senwang86** [comment]
> This UI looks fancy! Instead of "moving", I'd like to keep both (essentially two types of "widgets"). We need a "widget" system to allow users to configure what widgets to show in the sidebar and the bottom-status-bar.

Yeah, I recall this. The only concern is that if we need to maintain 2 sets of code for the same UI system, given that `Sidebar` and `BottomBar` have different space constraint.

**Comment by senwang86** [comment]
> A second thought: I probably want to hold on this, because the status items/buttons aren’t fixed, and I’m working on adding locally spawned runtimes to these sections. Working with sidebar seems easier to make changes.

SG. Just let me know once you finish the local runtime work.

**Comment by lihebi** [comment]
We'll have a widget system that allows users to configure "what" widgets to be displayed at "which" location, be it top or bottom.

**Comment by senwang86** [comment]
Screenshot 2023-08-31 at 6 56 17 PM

File: apps/ui/src/components/BottomAppBar.tsx [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    connected: { title: "connected", color: "green" },
+    connecting: { title: "connecting", color: "yellow" },
+    default: { title: "disconnected", color: "red" },
+  };
+  const wsStatus = wsStatusConfig[runtime.wsStatus || "default"];
+
+  return (
+    <ListItem key={runtimeId} disablePadding sx={{ margin: 0 }}>
+      {activeRuntime === runtimeId && (
+        <Tooltip title={wsStatus.title}>
```
Comment (lihebi) [inline apps/ui/src/components/BottomAppBar.tsx:98]: Crash when deleting a runtime.

Screenshot 2023-09-01 at 12 54 55 AM

Comment (senwang86) [reply apps/ui/src/components/BottomAppBar.tsx:98]: Is it tested after merged PR #501 ? it works fine on my end though.

**Review by lihebi — COMMENTED** [review]
I don't understand how c0b1fd071 creates Sidebar "widget". This commit looks like a few refactorings and a few style changes. It's better if you can create separate small PRs with focused changes.

**Comment by lihebi** [comment]
I used the bottom bar a little, and I feel that this design is a little over-engineered. The first principle of UI design IMO is to be **simple** and **explicit**.

1. Simple: I'm not a fan of replacing text with icons. Texts are simple and easy to understand.
2. Explicit: The bottom bar also hides multiple runtimes into 2nd-level menus, which is not explicit. We may very well need to see all runtime statuses, especially when we introduce dev/prod, training/inference, multiple languages.

Let's use plain text and refine the UI later.

**Comment by lihebi** [comment]
> we need to maintain 2 sets of code for the same UI system

Agree. This is also one of the main concerns that I hesitate to merge. I can feel that this will create headaches for future changes. Let's defer the UI polishing to the very end.

File: apps/ui/src/components/BottomAppBar.tsx [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+  const store = useContext(RepoContext)!;
+  const runtimeMap =
+    useStore(store, (state) => state.getRuntimeMap()) || new Map();
+  // Observe runtime change
+  const runtimeChanged = useStore(store, (state) => state.runtimeChanged);
+  // A dummy useEffect to indicate that runtimeChanged is used.
+  useEffect(() => {}, [runtimeChanged]);
+  const activeRuntime = useStore(store, (state) => state.activeRuntime);
+  const setActiveRuntime = useStore(store, (state) => state.setActiveRuntime);
+  const runtime = runtimeMap.get(runtimeId) || { wsStatus: "unknown" };
```
Comment (lihebi) [inline apps/ui/src/components/BottomAppBar.tsx:49]: [1/3] you set `runtime = {wsStatus: "uknonwn"}` as the fallback

File: apps/ui/src/components/BottomAppBar.tsx [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+  );
+
+  const runtimeStatusColors = { idle: "green", busy: "yellow", default: "red" };
+  const runtimeStatus = runtime.status ?? "default";
+  const wsStatusConfig = {
+    connected: { title: "connected", color: "green" },
+    connecting: { title: "connecting", color: "yellow" },
+    default: { title: "disconnected", color: "red" },
+  };
+  const wsStatus = wsStatusConfig[runtime.wsStatus || "default"];
```
Comment (lihebi) [inline apps/ui/src/components/BottomAppBar.tsx:93]: [2/3] And if that's `unknown`, this wsStatusConfig doesn't have the key, which leads to `wsStatus===undefined`

File: apps/ui/src/components/BottomAppBar.tsx (hunk 2 above)
Comment (lihebi) [reply apps/ui/src/components/BottomAppBar.tsx:98]: [3/3] And finally, you access `wsStatus.title` which crashes with "reading properties of undefined" error.

File: apps/ui/src/components/BottomAppBar.tsx (hunk 4 above)
Comment (lihebi) [reply apps/ui/src/components/BottomAppBar.tsx:93]: Adding some type annotations can help you catch such bugs early and more comprehensively.

**Comment by senwang86** [comment]
> Thanks @senwang86. My two cents: Can we move the status info to the top bar? The bottom of the screen is usually where the programmer's gaze is at: because code goes top to down. The status bar there sounds a distraction.

Hi @forrestbao , appreciate for the feedback! After some discussion, we agree that the status bar might not offer enough benefit, just as you suggested that it might cause users to be distracted. Thus it will not be merged.

**Pull request closed without merging**